sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -78,49 +84,105 @@
             });
     }
 
+    private static <T extends Resource> Mono<Pair<List<PartitionKeyRange>, QueryInfo>> getTargetRangesFromQueryPlan(
+        CosmosQueryRequestOptions cosmosQueryRequestOptions, DocumentCollection collection,
+        DefaultDocumentQueryExecutionContext<T> queryExecutionContext,
+        PartitionedQueryExecutionInfo partitionedQueryExecutionInfo, Instant planFetchStartTime,
+        Instant planFetchEndTime) {
+        QueryInfo queryInfo =
+            partitionedQueryExecutionInfo.getQueryInfo();
+        queryInfo.setQueryPlanDiagnosticsContext(new QueryInfo.QueryPlanDiagnosticsContext(planFetchStartTime,
+                                                                                           planFetchEndTime));
+        List<Range<String>> queryRanges =
+            partitionedQueryExecutionInfo.getQueryRanges();
+
+        if (isScopedToSinglePartition(cosmosQueryRequestOptions)) {
+            PartitionKeyInternal internalPartitionKey =
+                BridgeInternal.getPartitionKeyInternal(cosmosQueryRequestOptions.getPartitionKey());
+            Range<String> range = Range
+                                      .getPointRange(internalPartitionKey
+                                                         .getEffectivePartitionKeyString(internalPartitionKey,
+                                                                                         collection
+                                                                                             .getPartitionKey()));
+            queryRanges = Collections.singletonList(range);
+        }
+        return
+            queryExecutionContext.getTargetPartitionKeyRanges(collection.getResourceId(), queryRanges)
+                .map(pkRanges -> Pair.of(
+                    pkRanges,
+                    partitionedQueryExecutionInfo.getQueryInfo()));
+    }
+
+    private static void tryCacheQueryPlan(
+        SqlQuerySpec query,
+        PartitionedQueryExecutionInfo partitionedQueryExecutionInfo,
+        LRUCache<String, PartitionedQueryExecutionInfo> queryPlanCache) {
+        QueryInfo queryInfo = partitionedQueryExecutionInfo.getQueryInfo();
+        if (canCacheQuery(queryInfo) && !queryPlanCache.containsKey(query.getQueryText())) {
+            try {
+                queryPlanCache.put(query.getQueryText(), partitionedQueryExecutionInfo);
+            } catch (ConcurrentModificationException exception) {
+                logger.error("Error caching query plan: ", exception);
+            }
+        }
+    }
+
+    private static boolean canCacheQuery(QueryInfo queryInfo) {
+        // Query plan will not be cached for the types below
+        return !queryInfo.hasAggregates()
+                   && !queryInfo.hasDistinct()
+                   && !queryInfo.hasGroupBy()
+                   && !queryInfo.hasLimit()
+                   && !queryInfo.hasTop()
```

Comment:
Why would Top/Limit not work for single partition queries? Not even for relatively common model of Top 1?

Reply:
The query plan for these queries includes the hardcoded values for the top/limit/offset specified even if the query is parameterized, which makes the query plan tied to a single query